### PR TITLE
enh_s2t joint model

### PR DESCRIPTION
**Pull request by simpleoier**
Support Enhancement and [ASR, SLU, ST] downstream tasks.

Future PR (TODO)
- [ ] Pass the category info instead of using utterance ids
- [ ] Pre-trained models upload.
- [ ] MixIT

### CODE REVIEW

**Comment by sw005320** [comment]
@ftshijt, can you also review the part related to ST?

File: egs2/TEMPLATE/enh1/enh.sh [hunk 1]
```diff
@@ -201,6 +201,9 @@ fi
 [ -z "${valid_set}" ] &&   { log "${help_message}"; log "Error: --valid_set is required"  ; exit 2; };
 [ -z "${test_sets}" ] && { log "${help_message}"; log "Error: --test_sets is required"; exit 2; };
 
+# Extra files for enhancement process
+utt_extra_files="utt2category text utt2lang"
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh1/enh.sh:205]: Just a note: we need to introduce these files in the [speech enhancement documentation](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/README.md) in later PRs.

File: egs2/TEMPLATE/enh_asr1/scripts/audio/format_wav_scp.sh [hunk 2]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/audio/format_wav_scp.sh:1]: This file seems to be a copy of [egs2/TEMPLATE/asr1/scripts/audio/format_wav_scp.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/scripts/audio/format_wav_scp.sh).
It is better to make a symbolic link instead, like [egs2/TEMPLATE/enh1/scripts/audio](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/scripts/audio).

File: egs2/TEMPLATE/enh_asr1/scripts/audio/trim_silence.sh [hunk 3]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/audio/trim_silence.sh:1]: ditto

File: egs2/TEMPLATE/enh_asr1/scripts/feats/feat_to_shape.sh [hunk 4]
```diff
@@ -0,0 +1 @@
+../../../../../utils/feat_to_shape.sh
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/feats/feat_to_shape.sh:1]: It would be better to link the `feats` directory to [egs2/TEMPLATE/asr1/scripts/feats](https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/asr1/scripts/feats).
Because that would automatically reflect possible changes in the `asr1` template in the future.

File: egs2/TEMPLATE/enh_asr1/scripts/feats/make_fbank.sh [hunk 5]
```diff
@@ -0,0 +1 @@
+../../../../../utils/make_fbank.sh
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/feats/make_fbank.sh:1]: ditto

File: egs2/TEMPLATE/enh_asr1/scripts/feats/make_stft.sh [hunk 6]
```diff
@@ -0,0 +1 @@
+../../../../../utils/make_stft.sh
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/feats/make_stft.sh:1]: ditto

File: egs2/TEMPLATE/enh_asr1/scripts/utils/TEMPLATE_HF_Readme.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+@inproceedings{hayashi2020espnet,
+  title={{Espnet-TTS}: Unified, reproducible, and integratable open source end-to-end text-to-speech toolkit},
+  author={Hayashi, Tomoki and Yamamoto, Ryuichi and Inoue, Katsuki and Yoshimura, Takenori and Watanabe, Shinji and Toda, Tomoki and Takeda, Kazuya and Zhang, Yu and Tan, Xu},
+  booktitle={Proceedings of IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP)},
+  pages={7654--7658},
+  year={2020},
+  organization={IEEE}
+}';
+fi)
+\`\`\`
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/utils/TEMPLATE_HF_Readme.md:97]: It's better to keep these references consistent with those in the main README file: https://github.com/espnet/espnet#citations

Comment (Emrys365) [reply egs2/TEMPLATE/enh_asr1/scripts/utils/TEMPLATE_HF_Readme.md:97]: Maybe also add the latest work of ESPnet-SE++?

File: egs2/TEMPLATE/enh_asr1/scripts/utils/create_README_file.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                line = line.replace("<add_lang>", "en\n- zh\n- ja\n- multilingual")
+            else:
+                line = line.replace("<add_lang>", lang_name)
+        line = line.replace("<add_model_name>", model_name)
+        line = line.replace("<add_url>", url_name)
+        line = line.replace("<add_name>", user_name)
+        line = line.replace("<add_corpus>", corpus_name)
+        line = line.replace("<add_task_name>", task_name.upper())
+        line = line.replace("<add_recipe_task_name>", task_name.lower() + "1")
+        if "<add_tts_reference>" in line:
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/utils/create_README_file.py:48]: Why would `add_tts_reference` be used here?

File: egs2/TEMPLATE/enh_asr1/scripts/utils/get_model_names.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+from espnet_model_zoo.downloader import ModelDownloader
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/utils/get_model_names.py:1]: Better to replace with a symbolic link to [egs2/TEMPLATE/asr1/scripts/utils/get_model_names.py](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/scripts/utils/get_model_names.py).

File: egs2/TEMPLATE/enh_asr1/scripts/utils/perturb_data_dir_speed.sh [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        utils/apply_map.pl -f 1 "${destdir}"/utt_map <"${srcdir}"/${x} >"${destdir}"/${x}
+    fi
+done
+if [[ -f ${srcdir}/spk2gender ]]; then
+    utils/apply_map.pl -f 1 "${destdir}"/spk_map <"${srcdir}"/spk2gender >"${destdir}"/spk2gender
+fi
+
+rm "${destdir}"/spk_map "${destdir}"/utt_map "${destdir}"/reco_map 2>/dev/null
+echo "$0: generated speed-perturbed version of data in ${srcdir}, in ${destdir}"
+utils/fix_data_dir.sh "${destdir}"
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/utils/perturb_data_dir_speed.sh:118]: Maybe better to move the change to [egs2/TEMPLATE/asr1/scripts/utils/perturb_data_dir_speed.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/scripts/utils/perturb_data_dir_speed.sh) and make a symbolic link?

File: egs2/TEMPLATE/enh_asr1/scripts/utils/upload_models_to_hub.sh [hunk 11]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/scripts/utils/upload_models_to_hub.sh:1]: Better to replace with a symbolic link to [egs2/TEMPLATE/asr1/scripts/utils/upload_models_to_hub.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/scripts/utils/upload_models_to_hub.sh).

File: egs2/TEMPLATE/enh_asr1/setup.sh [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+# Copy
+for f in cmd.sh conf local; do
+    target="${dir}"/../../TEMPLATE/asr1/"${f}"
+    cp -r "${target}" "${dir}"
+    targets+="${dir}/${target} "
+done
+
+
+# Symlinks to TEMPLATE
+for f in asr.sh path.sh db.sh scripts pyscripts; do
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/setup.sh:44]: For the `scripts` directory, I think you need to link to `../../TEMPLATE/enh_asr1/scripts`.
Also, the `enh_asr.sh` should be linked.

For `asr.sh`, I am not sure whether we would need it.

You could take [egs2/TEMPLATE/enh1/setup.sh#L43-L48](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/setup.sh#L43-L48) as a reference.

File: egs2/TEMPLATE/enh_asr1/setup.sh [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+if [ ! -d "${dir}"/../../TEMPLATE ]; then
+    log "Error: ${dir}/../../TEMPLATE should exist. You may specify wrong directory."
+    exit 1
+fi
+
+targets=""
+
+# Copy
+for f in cmd.sh conf local; do
+    target="${dir}"/../../TEMPLATE/asr1/"${f}"
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/setup.sh:37]: Better to use `target="${dir}"/../../TEMPLATE/enh_asr1/"${f}"`. Otherwise, these newly added files/directories will not be used.

File: espnet2/bin/enh_s2t_train.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+from espnet2.tasks.enh_s2t import EnhS2TTask
+
+
+def get_parser():
+    parser = EnhS2TTask.get_parser()
+    return parser
+
+
+def main(cmd=None):
+    r"""ASR training.
```
Comment (Emrys365) [inline espnet2/bin/enh_s2t_train.py:11]: Replace `ASR` with `EnhS2T`?

File: espnet2/bin/enh_s2t_train.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+    Example:
+
+        % python enh_s2t_train.py asr --print_config --optim adadelta \
```
Comment (Emrys365) [inline espnet2/bin/enh_s2t_train.py:15]: What does `asr` mean here?

**Review by Emrys365 — COMMENTED** [review]
I just made a few general comments after a rough review. For some major changes, I will review them more carefully later.

File: espnet2/enh/espnet_enh_s2t_model.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+                text,
+                text_lengths,
+                src_text,
+                src_text_lengths,
+            )
+        else:
+            raise NotImplementedError(f"{type(self.s2t_model)} is not supported yet.")
+
+        if loss_enh is not None:
+            loss = loss_enh + loss_asr
```
Comment (LiChenda) [inline espnet2/enh/espnet_enh_s2t_model.py:176]: We may consider adding weight options for this MTL loss in the future.
e.g., `loss = loss_enh + lambda * loss_asr`

File: espnet2/enh/espnet_model.py [hunk 17]
```diff
@@ -136,6 +137,28 @@ def forward(
         speech_mix = speech_mix[:, : speech_lengths.max()]
 
         # model forward
+        speech_pre, feature_mix, feature_pre, others = self.forward_enhance(
```
Comment (LiChenda) [inline espnet2/enh/espnet_model.py:140]: Cool!

File: espnet2/tasks/enh.py [hunk 18]
```diff
@@ -233,12 +233,13 @@ def build_model(cls, args: argparse.Namespace) -> ESPnetEnhancementModel:
         decoder = decoder_choices.get_class(args.decoder)(**args.decoder_conf)
 
         loss_wrappers = []
-        for ctr in args.criterions:
-            criterion = criterion_choices.get_class(ctr["name"])(**ctr["conf"])
-            loss_wrapper = loss_wrapper_choices.get_class(ctr["wrapper"])(
-                criterion=criterion, **ctr["wrapper_conf"]
-            )
-            loss_wrappers.append(loss_wrapper)
+        if args.criterions is not None:
```
Comment (LiChenda) [inline espnet2/tasks/enh.py:236]: Thanks for adding this check!

**Review by LiChenda — COMMENTED** [review]
I also make some review comments for this PR.

File: espnet2/enh/espnet_enh_s2t_model.py (hunk 16 above)
Comment (simpleoier) [reply espnet2/enh/espnet_enh_s2t_model.py:176]: Yes. We will update it in the future.

File: egs2/TEMPLATE/enh1/enh.sh (hunk 1 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh1/enh.sh]: Cool. Just updated and added the description for utt2category.

File: egs2/TEMPLATE/enh_asr1/scripts/audio/format_wav_scp.sh (hunk 2 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/scripts/audio/format_wav_scp.sh]: Thanks. Have addressed this and the following similar comments.

File: egs2/TEMPLATE/enh_asr1/scripts/utils/TEMPLATE_HF_Readme.md (hunk 7 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/scripts/utils/TEMPLATE_HF_Readme.md]: We can do this when espnet-se++ is on arxiv or acceptet.

File: egs2/TEMPLATE/enh_asr1/scripts/utils/create_README_file.py (hunk 8 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/scripts/utils/create_README_file.py]: Now I linked the file from asr1. tts1 uses the same `create_README_file.py`. That's the reason I guess.

File: egs2/TEMPLATE/enh_asr1/setup.sh (hunk 12 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/setup.sh]: Cool. Have addressed it and the following.

File: espnet2/bin/enh_s2t_train.py (hunk 14 above)
Comment (simpleoier) [reply espnet2/bin/enh_s2t_train.py]: Done.

File: espnet2/bin/enh_s2t_train.py (hunk 15 above)
Comment (simpleoier) [reply espnet2/bin/enh_s2t_train.py]: fixed.

File: egs2/TEMPLATE/enh1/scripts/utils/perturb_enh_data_dir_speed.sh [hunk 19]
```diff
@@ -108,17 +111,16 @@ for scp_file in ${scp_files};do
   fi
 done
 
-if [[ -f ${srcdir}/text ]]; then
-    utils/apply_map.pl -f 1 "${destdir}"/utt_map <"${srcdir}"/text >"${destdir}"/text
-fi
+for x in text utt2lang ${utt_extra_files}; do
+    echo ${x}
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh1/scripts/utils/perturb_enh_data_dir_speed.sh]: Just suppressed it.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 20]
```diff
@@ -0,0 +1,1608 @@
...
+                                          # inference_enh_asr_model=valid.loss.ave.pth
+download_model= # Download a model from Model Zoo and use it for decoding.
+
+# Enhancement related arguments
+spk_num=1   # Number of speakers
+noise_type_num=1
+dereverb_ref_num=1
+# Evaluation related
+enh_inference_args="--normalize_output_wav true"
+scoring_protocol="STOI SDR SAR SIR SI_SNR PESQ"
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh]: Thanks for the info. removed it.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 21]
```diff
@@ -0,0 +1,1608 @@
...
+
+# Enh Training data related
+use_dereverb_ref=false
+use_noise_ref=false
+
+# Enhancement related
+dump_enh_wavs=true
+
+# Evaluation related
+scoring_protocol="STOI SDR SAR SIR SI_SNR"
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh]: Yes. Thanks for the checking.

File: espnet2/asr/frontend/s3prl.py [hunk 22]
```diff
@@ -86,10 +86,10 @@ def _get_upstream(self, frontend_conf):
```
Comment (sw005320) [inline espnet2/asr/frontend/s3prl.py:86]: Is it related to this PR?
If it takes time to review this PR, we may make this part a separate PR.

**Review by sw005320 — COMMENTED** [review]
- do we need to have a special inference code for enhanced speech translation and recognition? I thought that we may just need to have them for training since the inference interface (given a speech to output text) would not be changed w/ and w/o enhancement.

File: espnet2/asr/frontend/s3prl.py (hunk 22 above)
Comment (simpleoier) [reply espnet2/asr/frontend/s3prl.py:86]: > * do we need to have a special inference code for enhanced speech translation and recognition? I thought that we may just need to have them for training since the inference interface (given a speech to output text) would not be changed w/ and w/o enhancement.

The difference between `enh_s2t_inference` and `asr_inference` is that `enh_s2t_inference` can dump both the enhanced wavs and the text in the inference. I can remove the inference script if this is not necessary.



> Is it related to this PR? If it takes time to review this PR, we may make this part a separate PR.

In some sense, yes. We would like to backpropagate the gradient from ASR to Enh models.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 23]
```diff
@@ -0,0 +1,1604 @@
...
+num_splits_asr=1           # Number of splitting for lm corpus.
+
+# Upload model related
+hf_repo=
+
+# Decoding related
+use_k2=false      # Whether to use k2 based decoder
+batch_size=1
+inference_tag=    # Suffix to the result dir for decoding.
+inference_config= # Config for decoding.
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:102]: Similarly this also applies to `egs2/TEMPLATE/enh_st1/enh_st.sh`.

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 24]
```diff
@@ -0,0 +1,1843 @@
...
+                        --ref_channels "0" \
+                        --audio-format "${audio_format}" --fs "${fs}" ${_opts} \
+                        "data/${dset}/${spk}.scp" "${data_feats}${_suf}/${dset}" \
+                        "${data_feats}${_suf}/${dset}/logs/${spk}" "${data_feats}${_suf}/${dset}/data/${spk}"
+                done
+
+                echo "${feats_type}" > "${data_feats}${_suf}/${dset}/feats_type"
+            done
+
+        elif  [ "${feats_type}" = extracted ]; then
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:628]: What is the possible application of this case?

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 25]
```diff
@@ -0,0 +1,1843 @@
...
+            log "Stage 3: ${feats_type} extract: data/ -> ${data_feats}"
+            # Assumming you don't have wav.scp, but feats.scp is created by local/data.sh instead.
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:630]: typo: Assumming → Assuming

File: egs2/TEMPLATE/enh_st1/enh_st.sh (hunk 24 above)
Comment (Emrys365) [reply egs2/TEMPLATE/enh_st1/enh_st.sh:628]: If `feats_type` is not `raw`, will the cascade of enh and st models still work?
> I think there might be a mismatch in the feature extraction of the st module when using a non-`raw` `feats_type`.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 26]
```diff
@@ -0,0 +1,1604 @@
...
+                        "data/${dset}/${spk}.scp" "${data_feats}${_suf}/${dset}" \
+                        "${data_feats}${_suf}/${dset}/logs/${spk}" "${data_feats}${_suf}/${dset}/data/${spk}"
+
+                done
+
+                echo "${feats_type}" > "${data_feats}${_suf}/${dset}/feats_type"
+
+            done
+
+        elif  [ "${feats_type}" = extracted ]; then
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:561]: If `feats_type` is not `raw`, will the cascade of enh and asr models still work?
> I think there might be a mismatch in the feature extraction of the asr module when using a non-`raw` `feats_type`.

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 27]
```diff
@@ -0,0 +1,1843 @@
...
+                for protocol in ${scoring_protocol}; do
+                    # shellcheck disable=SC2046
+                    paste $(for j in $(seq ${spk_num}); do echo "${_dir}"/"${protocol}"_spk"${j}" ; done)  |
+                    awk 'BEGIN{sum=0}
+                        {n=0;score=0;for (i=2; i<=NF; i+=2){n+=1;score+=$i}; sum+=score/n}
+                        END{printf ("%.2f\n",sum/NR)}' > "${_dir}/result_${protocol,,}.txt"
+                done
+            done
+
+            ./scripts/utils/show_enh_score.sh "${_dir}/../.." > "${_dir}/../../RESULTS.md"
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:1746]:
```suggestion
            ./scripts/utils/show_enh_score.sh "${_dir}/../.." > "${_dir}/../../RESULTS_enh.md"
```

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 28]
```diff
@@ -0,0 +1,1843 @@
...
+        done
+        log "Evaluation result for observation: ${data_feats}/RESULTS.md"
+        log "Evaluation result for enhancement: ${enh_asr_exp}/enhanced/RESULTS.md"
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:1749]:
```suggestion
        log "Evaluation result for enhancement: ${enh_asr_exp}/RESULTS_enh.md"
```

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 29]
```diff
@@ -0,0 +1,1843 @@
...
+        if [ "${nlsyms_txt}" != none ]; then
+            _opts+="--option ${nlsyms_txt} "
+        fi
+        # shellcheck disable=SC2086
+        ${python} -m espnet2.bin.pack enh_s2t \
+            --enh_s2t_train_config "${enh_st_exp}"/config.yaml \
+            --enh_s2t_model_file "${enh_st_exp}"/"${inference_enh_st_model}" \
+            ${_opts} \
+            --option "${enh_st_exp}"/RESULTS.md \
+            --option "${enh_st_exp}"/RESULTS.md \
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:1785]:
```suggestion
            --option "${enh_st_exp}"/RESULTS_enh.md \
```

File: espnet2/tasks/enh_s2t.py [hunk 30]
```diff
@@ -0,0 +1,465 @@
...
+    # Add variable objects configurations
+    class_choices_list = [
+        # --enh_encoder and --enh_encoder_conf
+        enh_encoder_choices,
+        # --enh_separator and --enh_separator_conf
+        enh_separator_choices,
+        # --enh_decoder and --enh_decoder_conf
+        enh_decoder_choices,
+        # --enh_criterion and --enh_criterion_conf
+        enh_criterions_choices,
```
Comment (Emrys365) [inline espnet2/tasks/enh_s2t.py:166]: I guess this will only allow a single loss function for the ENH task. It is better to remove this from `class_choices_list` and add an argument `--criterions` in `add_task_arguments` as in [espnet2/tasks/enh.py](https://github.com/espnet/espnet/blob/master/espnet2/tasks/enh.py).

File: espnet2/tasks/enh_s2t.py [hunk 31]
```diff
@@ -0,0 +1,465 @@
...
+        else:
+            # Recognition mode
+            retval = ("speech",)
+        return retval
+
+    @classmethod
+    def optional_data_names(
+        cls, train: bool = True, inference: bool = False
+    ) -> Tuple[str, ...]:
+        retval = ["dereverb_ref"]
```
Comment (Emrys365) [inline espnet2/tasks/enh_s2t.py:418]:
```suggestion
        retval = ["dereverb_ref1"]
```

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 32]
```diff
@@ -0,0 +1,1657 @@
...
+inference_enh_asr_model=valid.acc.ave.pth # ASR model path for decoding.
+                                          # e.g.
+                                          # inference_enh_asr_model=train.loss.best.pth
+                                          # inference_enh_asr_model=3epoch.pth
+                                          # inference_enh_asr_model=valid.acc.best.pth
+                                          # inference_enh_asr_model=valid.loss.ave.pth
+download_model= # Download a model from Model Zoo and use it for decoding.
+
+# Enhancement related arguments
+spk_num=1   # Number of speakers
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_asr1/enh_asr.sh:117]: Better to add a warning if `spk_num` > 1 because we currently do not support it.

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 33]
```diff
@@ -0,0 +1,1843 @@
...
+inference_enh_st_model=valid.acc.ave.pth # ST model path for decoding.
+                                      # e.g.
+                                      # inference_enh_st_model=train.loss.best.pth
+                                      # inference_enh_st_model=3epoch.pth
+                                      # inference_enh_st_model=valid.acc.best.pth
+                                      # inference_enh_st_model=valid.loss.ave.pth
+download_model= # Download a model from Model Zoo and use it for decoding.
+
+# Enhancement related arguments
+spk_num=1   # Number of speakers
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:130]: Better to add a warning if `spk_num` > 1 because we currently do not support it.

File: espnet2/tasks/enh_s2t.py [hunk 34]
```diff
@@ -0,0 +1,465 @@
...
+        retval += ["speech_ref{}".format(n) for n in range(2, MAX_REFERENCE_NUM + 1)]
+        retval += ["text_ref{}".format(n) for n in range(2, MAX_REFERENCE_NUM + 1)]
```
Comment (Emrys365) [inline espnet2/tasks/enh_s2t.py:420]: Maybe remove this line, because`text_ref*` seems not prepared in `egs2/TEMPLATE/enh_asr1/enh_asr.sh` and `egs2/TEMPLATE/enh_st1/enh_st.sh`

File: espnet2/tasks/enh_s2t.py [hunk 35]
```diff
@@ -0,0 +1,465 @@
...
+                subtask_conf[attr] = (
+                    getattr(args, subtask + "_" + attr, None)
+                    if getattr(args, subtask + "_" + attr, None) is not None
+                    else getattr(args, attr, None)
+                )
+
+            if subtask in ["asr", "st"]:
+                m_subtask = "s2t"
+            else:
+                m_subtask = subtask
```
Comment (Emrys365) [inline espnet2/tasks/enh_s2t.py:448]: Is this for support of other Speech-to-Text tasks in the future? If so, it's better to add a warning if they are not supported now.

**Review by Emrys365 — COMMENTED** [review]
I finished my second-round review, except for two `s3prl` scripts as I am not familiar with them.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh (hunk 23 above)
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:102]: OK. I've added it.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 36]
```diff
@@ -0,0 +1,1604 @@
...
+
+                done
+
+                echo "${feats_type}" > "${data_feats}${_suf}/${dset}/feats_type"
+
+            done
+
+        elif  [ "${feats_type}" = extracted ]; then
+            log "Stage 3: ${feats_type} extract: data/ -> ${data_feats}"
+            # Assumming you don't have wav.scp, but feats.scp is created by local/data.sh instead.
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:563]: OK. Addressed.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh (hunk 26 above)
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:561]: I'll delete this part.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 37]
```diff
@@ -0,0 +1,1604 @@
...
+                for protocol in ${scoring_protocol}; do
+                    # shellcheck disable=SC2046
+                    paste $(for j in $(seq ${spk_num}); do echo "${_dir}"/"${protocol}"_spk"${j}" ; done)  |
+                    awk 'BEGIN{sum=0}
+                        {n=0;score=0;for (i=2; i<=NF; i+=2){n+=1;score+=$i}; sum+=score/n}
+                        END{printf ("%.2f\n",sum/NR)}' > "${_dir}/result_${protocol,,}.txt"
+                done
+            done
+
+            ./scripts/utils/show_enh_score.sh "${_dir}/../.." > "${_dir}/../../RESULTS.md"
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:1561]: Thanks!

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh (hunk 32 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh:117]: done

File: egs2/TEMPLATE/enh_st1/enh_st.sh (hunk 33 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_st1/enh_st.sh:130]: done

File: espnet2/tasks/enh_s2t.py (hunk 34 above)
Comment (simpleoier) [reply espnet2/tasks/enh_s2t.py:420]: OK

File: espnet2/tasks/enh_s2t.py (hunk 35 above)
Comment (simpleoier) [reply espnet2/tasks/enh_s2t.py:448]: OK. Actually, the subtask types are constrained in the argument definition. Anyway, I'll add a assertion error.

**Comment by simpleoier** [comment]
> I finished my second-round review, except for two `s3prl` scripts as I am not familiar with them.

Thanks. I think those s3prl files are fine.

File: egs2/TEMPLATE/enh_st1/enh_st.sh [hunk 38]
```diff
@@ -0,0 +1,1808 @@
...
+    --enh_st_exp           # Specify the directory path for ST experiment.
+                       # If this option is specified, enh_st_tag is ignored (default="${enh_st_exp}").
+    --enh_st_stats_dir     # Specify the directory path for ST statistics (default="${enh_st_stats_dir}").
+    --enh_st_config        # Config for st model training (default="${enh_st_config}").
+    --enh_st_args          # Arguments for st model training (default="${enh_st_args}").
+                           # e.g., --enh_st_args "--max_epoch 10"
+                           # Note that it will overwrite args in st config.
+    --pretrained_model=          # Pretrained model to load (default="${pretrained_model}").
+    --ignore_init_mismatch=      # Ignore mismatch parameter init with pretrained model (default="${ignore_init_mismatch}").
+    --pretrained_asr=          # Pretrained model to load (default="${pretrained_asr}").
```
Comment (Emrys365) [inline egs2/TEMPLATE/enh_st1/enh_st.sh:237]: Is this argument used?

Comment (simpleoier) [reply egs2/TEMPLATE/enh_st1/enh_st.sh]: removed.

File: egs2/TEMPLATE/enh_asr1/enh_asr.sh [hunk 39]
```diff
@@ -0,0 +1,1604 @@
...
+        if [ "${nlsyms_txt}" != none ]; then
+            _opts+="--option ${nlsyms_txt} "
+        fi
+        # shellcheck disable=SC2086
+        ${python} -m espnet2.bin.pack enh_s2t \
+            --enh_s2t_train_config "${enh_asr_exp}"/config.yaml \
+            --enh_s2t_model_file "${enh_asr_exp}"/"${inference_enh_asr_model}" \
+            ${_opts} \
+            --option "${enh_asr_exp}"/RESULTS.md \
+            --option "${enh_asr_exp}"/RESULTS.md \
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/TEMPLATE/enh_asr1/enh_asr.sh]: done.

File: egs2/TEMPLATE/enh_st1/enh_st.sh (hunk 29 above)
Comment (simpleoier) [reply egs2/TEMPLATE/enh_st1/enh_st.sh]: done.

File: espnet2/tasks/enh_s2t.py (hunk 30 above)
Comment (simpleoier) [reply espnet2/tasks/enh_s2t.py]: done.

**Review by ftshijt — APPROVED** [review]
It looks cool for me on the ST side. You have my pass for that part.

**Comment by sw005320** [comment]
OK for me as well.
@simpleoier, if you think this PR is ready, I’ll merge this PR.

**Comment by simpleoier** [comment]
Hi @sw005320 , it is ready. Thanks!

**Comment by Emrys365** [comment]
@simpleoier, I think we may also need to include the newly added tasks in the integration test: [ci/test_integration_espnet2.sh#L78](https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh#L78)

> To do that, some toy recipes need to be created in [`egs2/mini_an4`](https://github.com/espnet/espnet/tree/master/egs2/mini_an4).

**Comment by sw005320** [comment]
> @simpleoier, I think we may also need to include the newly added tasks in the integration test: [ci/test_integration_espnet2.sh#L78](https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh#L78)
> 
> > To do that, some toy recipes need to be created in [`egs2/mini_an4`](https://github.com/espnet/espnet/tree/master/egs2/mini_an4).

Good idea!

**Comment by simpleoier** [comment]
I'll work on the integration test. Thanks for the suggestion.
